integrations/transportation/ryanair.js

```diff
@@ -0,0 +1,141 @@
+// user info available at `${PROFILE_URL}${customerId}/profile/full/`
+
+async function logIn(username, password) {
+  const res = await agent
+    .post(LOGIN_URL)
+    .type('application/x-www-form-urlencoded')
+    .send({
+      password,
+      username,
+    });
+
+  if (!res.ok) {
+    throw HTTPError('Error while logging in.');
+  }
+
+  return {
+    token: res.body.token,
+    customerId: res.body.customerId,
+  };
+}
+
+async function connect(requestLogin, requestWebView) {
+  // Here we can request credentials etc..
+
+  // Here we can use two functions to invoke screens
+  // requestLogin() or requestWebView()
+  const { username, password } = await requestLogin();
+
+  if (!(password || '').length) {
+    throw HTTPError('Password cannot be empty');
+  }
+
+  return logIn(username, password);
+}
+
+function disconnect() {
+  // Here we should do any cleanup (deleting tokens etc..)
+  return {};
+}
+
+async function getAllFlights(bookings, customerId, token) {
+  const allFlights = [];
+  try {
+    await Promise.all(bookings.map(async (entry) => {
+      await agent
+        .put(`${PROFILE_URL}${customerId}/bookings/booking`)
+        .set('Accept', 'application/json')
+        .set('X-Auth-Token', token)
+        .send({
+          bookingId: entry.bookingId,
+          pnr: entry.pnr,
+        })
+        .then((res) => {
+          if (res.ok) {
+            res.body.Flights.forEach((singleFlight) => {
+              allFlights.push({
+                bookingId: res.body.BookingId,
+                flightInfo: singleFlight,
+                pnr: res.body.RecordLocator,
+              });
+            });
+          }
+        });
+    }));
+  } catch (error) {
+    throw HTTPError(`Error while fetching past flights. ${error}`);
+  }
+
+  return allFlights;
+}
+
+async function getPastBookings(customerId, token) {
+  const pastBookings = await agent
+    .get(`${PROFILE_URL}${customerId}/bookings/past`)
+    .set('Accept', 'application/json')
+    .set('X-Auth-Token', token);
+
+  if (!pastBookings.ok) {
+    throw HTTPError(`Error while fetching past bookings. ${pastBookings}`);
+  }
+
+  return pastBookings.body.bookings.filter(entry => entry.status === 'Confirmed');
+}
+
+async function getActivities(pastBookings, customerId, token) {
+  // WHY: there can be multiple flights in one booking
+  const allFlights = await getAllFlights(pastBookings, customerId, token);
+
+  // WHY: there can be multiple bookings for the same flight (e.g. for different passengers)
+  // HOW: filters flights with the same flight number
+  const uniqueFlights = Array.from(new Set(allFlights.map(a => a.flightInfo.FlightNumber)))
+    .map(mappedFlightNumber => allFlights.find(a => a.flightInfo.FlightNumber === mappedFlightNumber));
+
+  const activities = Object.values(uniqueFlights)
+    .map(k => ({
+      // TODO: make sure it's always unique
+      id: `B${k.bookingId}${k.flightInfo.FlightNumber}PNR${k.pnr}`,
```

Comment:
isn't the booking id sufficient?
Here, I would prefix with "ryanair" to make sure that it stays unique across integrations

Reply:
nope, flight number is also needed as there can be multiple different flights in one booking